Header to metadata filter (#2436)

rgs1 opened the pull request:
Add support for extracting dynamic metadata from requests. This can then
be used as static metadata would be used (e.g.: for subset load balancer
metadata matches, logging, etc).

Risk Level: Low

Testing: unit-test

Docs Changes:
Basic docs, more later.

Release Notes:
N/A

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

bmetzdorf commented:
Hi @rgs1 ! Very nice! Let me know when you think it's ready to be tested. I see you're still adding commits. I'm happy to test it and confirm that it works for me too.

rgs1 commented:
@bmetzdorf yep sorry, still ironing out a few things... Will let you know!

rgs1 commented:
@bmetzdorf @ggreenway ok, I think it's going through now... Sorry for the noise. Let me know if this goes along with what you are thinking. Thanks! 

cc: @zuercher who filed #2436.

rgs1 commented:
@bmetzdorf sorry, had to rebase plus i added one more commit to avoid string copies.

rgs1 commented:
Btw, I am not sold on calling this filter `header-to-metadata`... That's just what we called it internally. Something more descriptive would be better.

bmetzdorf commented:
Hi @rgs1.

So, the simple request_rules use case that I'm interested in works fine for plain HTTP, but as you already mentioned in #2436 it is missing regex extraction support. In addition to that though, it seems as if currently (and I know there's some work in progress to overhaul this) this (or any) http filter does not get called for *websocket* connections, which unfortunately is what I'd require (sigh..).

So I'm wondering if we could provide some functionality of this in the route matcher (which does work for websocket connections) instead:

```
    - match:
        prefix: "/"
        headers:
        - name: x-something
          value: abc([0-9]+)
          regex: true
      route:
        use_websocket: true
        cluster: mycluster
        metadata_match:
          filter_metadata:
            envoy.lb:
              digits_from_x_something: \1
```
We could also have two different implementations, until the websocket stuff gets refactored.

What do you think?

rgs1 commented:
@bmetzdorf we've got many many route matchers, so if we implement it there we'd need to be able to configure this at the `RouteConfiguration` [0] or `Route` [1] level — to avoid the copy/pasta. But yeah, I think it would work for us.

cc: @derekargueta 

[0] https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/rds.proto#L39
[1] https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/route/route.proto#L116

bmetzdorf commented:
@rgs1 If you have many route matchers, would the `Route` level still be ok for you?

Long term I'd think that the HTTP filter is the right approach. The filter should work with the upcoming websocket changes. Until then the limited functionality I require could be provided in the `Route` level, which would be referencing a regex capture group inside `metadata_match`. Regex matching is already supported.

rgs1 commented:
@bmetzdorf yeah, `Route` level is good I think.

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.h:
```diff
@@ -0,0 +1,141 @@
+namespace HttpFilters {
+namespace HeaderToMetadataFilter {
+
+enum class MetadataType { String, Number };
```
Seems like most of these enums, structs or typedefs are not something you want to publicly expose outside your filter, so they should probably become private members of the Config class. Also, consider just using the Protobuf types for MetadataKeyValue and MetadataType. I don't think the conversion buys you much.

rgs1 replied:
I am still undecided about avoiding the conversion from Protobuf
messages to internal structs, mostly because of the LowerCaseString
header field that I'd like to avoid converting/creating on every
request. So there's a small win there — we might also want to extend
things a bit more of which point the conversions might be useful. I
did not make the Rule related structs private because they are shared
between Config and the Filter class.. I could make them private to
Config and make Filter a friend of class but at that point I am
not sure if pays off anymore. Thoughts?

zuercher replied:
I agree that it's worth keeping the LowerCaseString. You could store the protobuf Rule objects directly along side their LowerCaseString names as a `std::vector<std::pair<Http::LowerCaseString, envoy::config::filter::http::header_to_metadata::v2::Config::Rule>>` or with a struct that has the LowerCaseString and the protobuf Rule as fields.

In the future, if there are other pre-computed structures necessary, I think it makes sense to start introducing extra classes.

rgs1 commented:
@zuercher thanks for the review! I've addressed most of your comments in 471e25fc5163596a31829648b8a488344a83e855, the commit description has details and questions about the things I am not so sure about... LMK what you think. Thanks!

rgs1 commented on api/envoy/config/filter/http/header_to_metadata/v2/header_to_metadata.proto:
```diff
@@ -0,0 +1,60 @@
+    //
+    // The value field is ignored for this case, since the header value
+    // will be used.
+    KeyValuePair on_header_present = 2;
```
you are correct, thanks for catching this.

zuercher replied:
Doesn't the second sentence on the on_header_present docs still need revision (or deletion)?

rgs1 replied:
@zuercher oops — fixed in the last commit.

rgs1 commented:
@zuercher i addressed your latest comments, lmk if i missed anything — thanks!

rgs1 commented:
Hmm, we should probably emit some stats (e.g.: total count of  metadata keys added, total per key, etc)? We haven't had a need for them internally (yet), but I actually think they could be useful for users to validate things are working as expected. Thoughts?

rgs1 commented:
@zuercher lmk if you think we should have stats for this iteration... or maybe that's something that can be addressed later?

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -0,0 +1,173 @@
+
+      // Key empty means keyval wasn't set.
+      // TODO(rgs): is there a more idiomatic way to do this?
+      if (keyval.key().empty()) {
```
If you skip this block when there's no on_header_missing struct, I think the validation makes it so this will never happen, and you can remove the check.

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -0,0 +1,173 @@
+      if (rule.remove()) {
+        headers.remove(header);
+      }
+    } else {
```
Should this be `else if (rule.has_on_header_missing())` to handle the case where no on_header_missing value is given?

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -0,0 +1,173 @@
+    const Http::HeaderEntry* header_entry = headers.get(header);
+
+    if (header_entry != nullptr) {
+      const auto& keyval = rule.on_header_present();
```
I think something needs to validate that on_header_present was actually set. I believe this can be done with annotations on the protobuf, but I'm not sure of the exact syntax.

zuercher commented:
@rgs1 I think you could leave stats for another pass. It'll keep this simpler and then the stats change will be easy to review as well.

rgs1 commented:
@zuercher addressed, thanks! Couldn't find the annotation to ensure that at least one of `on_header_present` or `on_header_missing` is set. I only see `oneof`:

https://developers.google.com/protocol-buffers/docs/proto#oneof

I'll keep looking. In the meanwhile, it's handled via code — we should probably also 
log an error if we find a rule without any action (on present or on missing).

rgs1 commented:
@zuercher we could change the protobuf defs to make it easier to validate things at that level. If we reworked the `Config` message to look something like:

```
message Config {
  enum ValueType {
    STRING = 0;
    NUMBER = 1;
  }

  // Trigger an action if a header is either present or missing.
  enum HeaderTrigger {
    present = 0;
    missing = 1;
  }

  message Action {
    // What triggers this action.
    HeaderTrigger trigger = 1 [(validate.rules).enum.defined_only = true];

    // The namespace — if this is empty, the filter's namespace will be used.
    string metadata_namespace = 2;

    // The key to use within the namespace.
    string key = 3 [(validate.rules).string.min_bytes = 1];

    // The value to pair with the given key.
    //
    // When used with `HeaderTrigger.present`, if value is non-empty it'll be used
    // instead of the header value. If both are empty, no metadata is added.
    //
    // When used with `HeaderTrigger.missing`, a non-empty value must be provided
    // otherwise no metadata is added.
    string value = 4;

    // The value's type — defaults to string.
    ValueType type = 5;
  }

  // Apply to requests or replies?
  enum RuleType {
    requests = 0;
    replies = 1;
  }

  // A Rule defines what metadata to apply when a header is present or missing.
  message Rule {
    // What to apply this rule to.
    RuleType type = 1 [(validate.rules).enum.defined_only = true];

    // The header that triggers this rule — required.
    string header = 2 [(validate.rules).string.min_bytes = 1];

    // The action to apply to this header.
    repeated Action actions = 3 [(validate.rules).repeated .min_items = 1];

    // Whether or not to remove the header after a rule is applied.
    //
    // This prevents headers from leaking.
    bool remove = 4;
  }

  // The list of rules to apply to requests/responses.
  repeated Rule rules = 1 [(validate.rules).repeated .min_items = 1];
}
```

it would make it easier to express validations.... Thoughts?

zuercher commented:
I think the version that feels most natural to configure is probably the version to use, as long as the validation occurs somewhere.

rgs1 commented:
@zuercher lets go with what's there then. I've pushed a commit that rejects `Config` objects with rules that have neither `on_header_present` nor `on_header_missing` set, lmk what you think. Thanks!

rgs1 commented:
I _think_ the `h1_capture_fuzz_test` failure in unrelated... will look in a bit.

rgs1 commented:
Lets see if master fixes the `h1_capture_fuzz_test` failure...

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -0,0 +1,184 @@
+  encoder_callbacks_ = &callbacks;
+}
+
+bool HeaderToMetadataFilter::addMetadata(StructMap& map, const std::string& metaNamespace,
```
Sorry I missed this before, but our variable style is all-lowercase/snake_case, so "meta_namespace".

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -0,0 +1,184 @@
+  }
+
+  // Have we seen this namespace before?
+  auto namespaceIter = map.find(metaNamespace);
```
Likewise, "namespace_iter"

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -0,0 +1,184 @@
+void HeaderToMetadataFilter::writeHeaderToMetadata(Http::HeaderMap& headers,
+                                                   const HeaderToMetadataRules& rules,
+                                                   Http::StreamFilterCallbacks& callbacks) {
+  StructMap structsByNamespace;
```
"structs_by_namespace"

zuercher commented on source/extensions/filters/http/header_to_metadata/header_to_metadata_filter.cc:
```diff
@@ -0,0 +1,184 @@
+}
+
+bool Config::configToVector(const ProtobufRepeatedRule& protoRules, HeaderToMetadataRules& vector) {
+  if (protoRules.size() == 0) {
```
proto_rules

rgs1 commented:
@zuercher thanks for the review!

@ggreenway @mattklein123 @alyssawilk any of you with cycles to review this filter?

alyssawilk commented:
I think per https://github.com/envoyproxy/envoy/blob/master/GOVERNANCE.md#extension-addition-policy
If you add a CODEOWNERS with you and zuercher (assuming he volunteers to own) you're good to go with his approval.  And fixed format :-)

mattklein123 commented:
Yup per @alyssawilk that's fine with me. I don't have time to review this right now. Please make sure that all the docs are in place though. Thank you!!

rgs1 commented:
@alyssawilk ah nice, will do!

@zuercher are you ok with being added to CODEOWNERS for this? :-) (no pressure)

zuercher commented on api/envoy/config/filter/http/header_to_metadata/v2/header_to_metadata.proto:
```diff
@@ -0,0 +1,69 @@
+
+import "validate/validate.proto";
+
+// [#protodoc-title: Header-To-Metadata Fitlter]
```
Filter is misspelled.

zuercher commented on api/envoy/config/filter/http/header_to_metadata/v2/header_to_metadata.proto:
```diff
@@ -0,0 +1,69 @@
+// The configuration for transforming headers into metadata. This is useful
+// for matching load balancer subsets, logging, etc.
+//
+// :ref:`configuration overview <config_http_filters_header_to_metadata>`.
```
This looks weird in the generated docs (https://65061-65214191-gh.circle-artifacts.com/0/source/generated/docs/api-v2/config/filter/http/header_to_metadata/v2/header_to_metadata.proto.html). Should probably have "Header-to-Metadata" at before the link so it reads "Header-to-metadata configuration overrview."

rgs1 commented:
@zuercher done — i went with `Header to Metadata` though, since the `IPTagging` filter also interchanges `IP Tagging` and `IPTagging`.... though I'll switch back to `Header-To-Metadata` if you have strong feelings about it :-) 

Output is here: https://65121-65214191-gh.circle-artifacts.com/0/source/generated/docs/api-v2/config/filter/http/header_to_metadata/v2/header_to_metadata.proto.html

bmetzdorf commented:
Hi @rgs1 . Unfortunately my use case requires websockets with early body, which violates the websocket spec. While #3301 / #3376 might still allow this for now (I haven't tested it yet) it will be a problem in the future when strict websocket support will be enforced, in which case the http_filter approach for extracting metadata won't work for me. @alyssawilk @ggreenway Do you see any chance that we might allow early websocket bodies even in the long term?

alyssawilk commented:
We've got regression tests for early websocket bodies for both the old style and new style websockets so hopefully you'll be set.

When you say "websockets with early body" do you mean explicit content-length != 0, or websocket payload arriving with the upgrade request?  I *could* see Envoy rejecting GET-with-body but I don't think anyone plans on killing off early data handling, especially if you comment on that regression test that there are users using it.

bmetzdorf commented:
@alyssawilk I use both, upgrade request coming in with body payload as well as a upgrade response with content-length > 0. Is that something we could entertain to support long term (maybe hidden behind a `use_relaxed_websockets` if needed)?

alyssawilk commented:
More likely a more generic allow_disallowed_bodies or something but I suspect we could leave it config available.   As I'd likely be the person implementing any adhere-to-the-spec feature and @mattklein123 or I would likely be the ones to review if someone else writes it, we'll keep your traffic in mind :-)

bmetzdorf commented:
That sounds great! Thanks @alyssawilk !

bmetzdorf commented:
@rgs1 In regards to regex, what do you think about something like this?

```
http_filters:
  - name: envoy.filters.http.header_to_metadata
    config:
      request_rules:
        - header: x-version
          match: "^[0-9]+\.([0-9]+)$"
          on_header_present:
            metadata_namespace: envoy.lb
            key: version
            type: STRING
          on_header_match:
            metadata_namespace: envoy.lb
            key: minor_version
            value: "$1"
            type: STRING
          on_header_missing:
            metadata_namespace: envoy.lb
            key: default
            value: 'true'
            type: STRING
          remove: false
```

rgs1 commented:
@bmetzdorf generally lgtm, lets try to make it consistent with how exact match vs regex match is implemented in other places... Something like this perhaps:

https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/route/route.proto#L941

What do you think?

bmetzdorf commented:
@rgs1 I think `regex_match` and `invert_match` (to not have to use negative lookahead / lookbehind) would be sufficient for now. Should we model `on_header_present` with `present_match` instead and `on_header_missing` with `present_match + invert_match` and copy/re-use route.HeaderMatcher (see https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/route/route.proto.html#envoy-api-msg-route-headermatcher)?

Examples:

Now:
```
http_filters:
  - name: envoy.filters.http.header_to_metadata
    config:
      request_rules:
        - header: x-version
          on_header_present:
            metadata_namespace: envoy.lb
            key: version
            type: STRING
          on_header_missing:
            metadata_namespace: envoy.lb
            key: default
            value: 'true'
            type: STRING
          remove: false
```

Future:
```
http_filters:
  - name: envoy.filters.http.header_to_metadata
    config:
      request_rules:
        headers:
          - name: x-version
            present_match: true
            metadata_namespace: envoy.lb
            key: version
            type: STRING
          - name: x-version
            present_match: true
            invert_match: true
            metadata_namespace: envoy.lb
            key: default
            value: 'true'
            type: STRING
          - name: x-version
            regex_match: "^[0-9]+\.([0-9]+)$"
            metadata_namespace: envoy.lb
            key: default
            value: "$1"
            type: STRING
```

with `value` being optional and defaulting to the actual complete header value (unless header is missing, then defaulting to empty string)?

rgs1 commented:
@bmetzdorf I like the idea of reusing/emulating `route.HeaderMatcher`... However I am not sure what the policy is with breaking config for existing filters.... @zuercher ?

rgs1 commented:
Yeah that would work.

rgs1 commented:
@bmetzdorf that looks good

bmetzdorf commented:
@rgs1 While thinking about this more I'm realizing that it might be easier to create a new filter and retire the current one later. I think it could be a little bit messy trying to cram both types of configuration and logic into the existing one. What do you think?

rgs1 commented:
@bmetzdorf sure that's fine too
